docs/python-grader/index.md

```diff
@@ -0,0 +1,203 @@
+# Python Autograder
+
+This file documents the default Python autograder included in the `prairielearn/grader-python` Docker image.  For general information on how to set up an external grader, visit the [external grading](../externalGrading.md) page. 
+
+## Setting up
+
+### `info.json`
+
+The question should be first set up to enable [external grading](../externalGrading.md), with `"gradingMethod": "External"` set in the `info.json` settings.  To use the specific Python autograder detailed in this document, `"image"` should be set to `"prairielearn/grader-python"` and `"entrypoint"` should point to `"/python_autograder/run.sh"` in the `"externalGradingOptions"` dictionary.
+
+A full `info.json` file should look something like:
+
+```javascript
+{
+    "uuid": "...",
+    "title": "...",
+    "topic": "...",
+    "tags": [...],
+    "type": "v3",
+    "singleVariant": true,
+    "gradingMethod": "External",
+    "externalGradingOptions": {
+        "enabled": true,
+        "image": "prairielearn/grader-python",
+        "entrypoint": "/python_autograder/run.sh",
+    }
+}
+```
+
+### `server.py`
+
+The server code in the `generate()` function must define the list of variables or functions that will be passed to the autograded student code as `names_for_user`, and also those that will be passed from the student code to the test code as `names_from_user`.  Only variables or functions listed in `names_for_user` will be accessible by the user from the setup code; only names listed in `names_from_user` will be accessible by the test cases from the user code.
+
+These are stored as a list of dictionary objects in the `data["params"]` dict.  The above `names_for_user` and `names_from_user` lists are stored as separate keys in `params`. Each variable dictionary has the following format:
+
+```json
+{
+    "name": "(name of the variable)",
+    "description": "(Human readable description of the variable)",
+    "type": "(Human readable type of the variable)"
+}
+```
+
+A full example is included here, taken from the question `demoAutograderSquare`:
+
+```python
+def generate(data):
+    data["params"]["names_for_user"] = [
+        {"name": "x", "description": "Description of the variable", "type": "float"},
+    ]
+    data["params"]["names_from_user"] = [
+        {"name": "x_sq", "description": "The square of $x$", "type": "float"},
+    ]
+```
```

Comment:
````suggestion
These are stored as a list of dictionary objects in the `data["params"]` dict.  The above `names_for_user` and `names_from_user` lists are stored as separate keys in `params`. For example:

```python
def generate(data):
    data["params"]["names_for_user"] = [
        {"name": "x", "description": "Description of the variable", "type": "float"},
    ]
    data["params"]["names_from_user"] = [
        {"name": "x_sq", "description": "The square of $x$", "type": "float"},
    ]
```

Each variable dictionary has entries `name` (the Python variable name in the code), `description` (human readable), and `type` (human readable). These variable lists are used for two purposes: (1) showing students which variables are used, and (2) making variables available to the student code and autograder code.
````